blog/introduccion-typescript.md

```diff
@@ -0,0 +1,261 @@
+
+### Cuál es el valor agregado de TypeScript?
+
+Ya entendimos que una de sus funcionalidades es actuar como un transpilador, pero si sólo hiciera eso, podríamos seguir utilizando babel.
+
+Fundamentalmente, TypeScript le incorpora dos nuevas características a JavaScript:
+
+* Chequeo estático de tipos
+* Annotations (ya lo veremos en próximos posts)
+
+## TypeScript en acción
+
+JavaScript es un lenguage débilmente tipado. Esto lo podemos observar de la siguiente manera:
+
+```javascript
+> typeof 2
+< "number"
+> typeof "hola"
+< "string"
+> typeof {que: "tal"}
+< "object"
+```
+
+JavaScript entiende que 2 es un número y "hola" un string, pero al no ser un lenguaje fuertemente tipado, no hay nada que nos prohiba hacer esto:
+
+```javascript
+var miVariable = 2;
+miVariable = "hola";
+
+// o cosas como
+
+fs.readFile(9, function(err, buffer) {
+  // ...
+});
+```
+
+Probablemente, no sea lo que queríamos hacer, y muchas veces resultará en **errores silenciosos**.
+
+Allí es donde los tipos nos pueden ayudar. Los tipos son **contratos** y de la misma manera que necesitamos leyes (contratos sociales) para no vivir en anarquía, los tipos nos permiten mantener la paz en nuestro código.
+
+Veamos otro ejemplo:
+
+```javascript
+function sumar(a, b) {
+  return a + b;
+}
+```
+
+Por lo que vimos en los ejemplos anteriores, nada nos impediría ejecutar lo siguiente:
+
+```javascript
+sumar("Nombre-", "Apellido");
+> "Nombre-Apellido"
+
+sumar("1","2");
+> "12"
+```
+
+Históricamente tratamos de resolver el inconveniente anterior con **contratos** dentro de los comentarios. En este caso, queremos aclarar la intención de la función es sumar números.
+
+```javascript
+/* Suma dos numeros
+ * @param number a : el primer numero
+ * @param number b : el segundo numero
+ * @return number : la suma del primero con el segundo
+ */
+function sumar(a, b) {
+  return a + b;
+}
+```
+Pero este tipo de contrato es muy débil. Es como tener la ley pero sin la 👮‍ que nos ayude a garantizar el orden. Necesitamos TypeScript!
+
+Veamos como quedaría el ejemplo anterior utilizando TypeScript
+
+```javascript
+function sumar(a: number, b: number): number {
+  return a + b;
+}
+```
+
+Si intentaramos sumar con valores no numéricos, TypeScript nos avisará que no estamos cumpliendo con las reglas establecidas! Además, nuestro código queda auto-documentado, no necesitamos más comentarios.
+
+Pueden ver el ejemplo haciendo click [acá](<http://www.typescriptlang.org/play/index.html#src=function%20sumar(a%3A%20number%2C%20b%3A%20number)%3A%20number%20%7B%0D%0A%20%20return%20a%20%2B%20b%3B%0D%0A%7D%0D%0A%0D%0A%2F%2Fdescomenten%20esto!%20sumar(%221%22%2C2)>)
+
+### La madre de todos los linters 
+Desde la versión 2.3, se puede agregar la opción @ts-check. Con esto, TypeScript no sólo hará un chequeo de tipos, sino que también, funcionará como linter!
```

Comment:
Fijate si podes cambiar este texto con lo que hablamos la otra vez, de que lo del @ts-check es para archivos javascript que todavía no quieren ser convertidos a typescript por alguna razón.

Osea, uno puede pensar que typescript es como un transpilador porque "traduce" el codigo ES6,7 para atrás y además como un linter / policía que se queja cuando alguna regla está mal. Y el tema de mother of all linters es porque un linter como jslint tiene reglas muy simples, como que termines con punto y coma, que pongas un espacio. El "linter" que te da typescript es mucho más fuerte, tiene que ver con la semantica de tu código, sabe lo que reciben y lo que devuelven tus funciones y tiene reglas mucho más complejas, que al igual que un linter, están alli para mejorar la calidad de tu código.

Otra vez, typescript por estar tipado se puede pensar como un super linter, pero además, desde la versión 2.3 que también lo podes usar como un linter posta, porque podes usarlo con tus archivos ESX verdaderos (no un "lenguaje nuevo") e igual te va a dar información copada.

Si queres, para demostrar este punto, podes sacarle un screenshot al VSCode con alguna función tipada y poner con el mouse encima de algún parámetro y ver como el intelisense de typescript te muestra inferencia de tipos por ejemplo, o algún error.

Si necesitas armar un proyecto podes

- Partir del typescript-node-starter
- Crear un archivo `.js` dentro del src
- Agregarle lo de @ts-check
- Modificar el tsconfig para que tenga [allowJs y/o checkJs](https://www.typescriptlang.org/docs/handbook/compiler-options.html)

Podes usar un código como este
```
const n = 1; // Typescript deberia inferir que es un número
/* Suma dos numeros
  * @param number a : el primer numero
  * @param number b : el segundo numero
  * @return number : la suma del primero con el segundo
  */
 function sumar(a, b) {
   return a + b;
 }
```